File: trunplay-backend/src/api/plans.py
"""
Plans API endpoints.
"""
from typing import List
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session

from ..database.models import get_db
from ..database import crud, schemas
from ..services.scheduler import get_scheduler
from ..services.playback import get_playback_service

router = APIRouter(prefix="/plans", tags=["plans"])


@router.get("", response_model=List[schemas.PlanResponse])
def get_plans(skip: int = 0, limit: int = 100, db: Session = Depends(get_db)):
    """Get all plans."""
    plans = crud.get_plans(db, skip=skip, limit=limit)
    return plans


@router.get("/{plan_id}", response_model=schemas.PlanResponse)
def get_plan(plan_id: str, db: Session = Depends(get_db)):
    """Get a specific plan."""
    plan = crud.get_plan(db, plan_id)
    if not plan:
        raise HTTPException(status_code=404, detail="Plan not found")
    return plan


@router.post("", response_model=schemas.PlanResponse)
def create_plan(plan: schemas.PlanCreate, db: Session = Depends(get_db)):
    """Create a new plan."""
    db_plan = crud.create_plan(db, plan)

    # Schedule if active
    if db_plan.is_active:
        scheduler = get_scheduler()
        scheduler.schedule_plan(
            plan_id=db_plan.id,
            plan_title=db_plan.title,
            start_time=db_plan.start_time,
            repeat_days=db_plan.repeat_days,
            skip_holidays=bool(db_plan.skip_holidays)
        )

    return db_plan


@router.put("/{plan_id}", response_model=schemas.PlanResponse)
def update_plan(plan_id: str, plan: schemas.PlanUpdate, db: Session = Depends(get_db)):
    """Update a plan."""
    db_plan = crud.update_plan(db, plan_id, plan)
    if not db_plan:
        raise HTTPException(status_code=404, detail="Plan not found")

    # Update schedule
    scheduler = get_scheduler()
    if db_plan.is_active:
        scheduler.schedule_plan(
            plan_id=db_plan.id,
            plan_title=db_plan.title,
            start_time=db_plan.start_time,
            repeat_days=db_plan.repeat_days,
            skip_holidays=bool(db_plan.skip_holidays)
        )
    else:
        scheduler.cancel_plan(plan_id)

    return db_plan


@router.delete("/{plan_id}")
def delete_plan(plan_id: str, db: Session = Depends(get_db)):
    """Delete a plan."""
    # Cancel schedule first
    scheduler = get_scheduler()
    scheduler.cancel_plan(plan_id)

    success = crud.delete_plan(db, plan_id)
    if not success:
        raise HTTPException(status_code=404, detail="Plan not found")

    return {"message": "Plan deleted"}


@router.post("/{plan_id}/activate", response_model=schemas.PlanResponse)
def activate_plan(plan_id: str, db: Session = Depends(get_db)):
    """Activate a plan."""
    plan = crud.get_plan(db, plan_id)
    if not plan:
        raise HTTPException(status_code=404, detail="Plan not found")

    plan = crud.update_plan(db, plan_id, schemas.PlanUpdate(is_active=True))

    # Schedule
    scheduler = get_scheduler()
    scheduler.schedule_plan(
        plan_id=plan.id,
        plan_title=plan.title,
        start_time=plan.start_time,
        repeat_days=plan.repeat_days,
        skip_holidays=bool(plan.skip_holidays)
    )

    return plan


@router.post("/{plan_id}/deactivate", response_model=schemas.PlanResponse)
def deactivate_plan(plan_id: str, db: Session = Depends(get_db)):
    """Deactivate a plan."""
    plan = crud.update_plan(db, plan_id, schemas.PlanUpdate(is_active=False))
    if not plan:
        raise HTTPException(status_code=404, detail="Plan not found")

    # Cancel schedule
    scheduler = get_scheduler()
    scheduler.cancel_plan(plan_id)

    return plan


@router.post("/{plan_id}/play")
async def play_plan(
    plan_id: str,
    start_position: int = 0,
    db: Session = Depends(get_db)
):
    """Immediately play a plan."""
    plan = crud.get_plan(db, plan_id)
    if not plan:
        raise HTTPException(status_code=404, detail="Plan not found")

    playback_service = get_playback_service()
    success = await playback_service.start_playback(
        db=db,
        plan=plan,
        trigger_type=schemas.TriggerType.MANUAL,
        start_position=start_position
    )

    if not success:
        raise HTTPException(status_code=500, detail="Failed to start playback")

    return {"message": "Playback started", "plan_id": plan_id}


@router.post("/{plan_id}/reset-progress")
def reset_plan_progress(plan_id: str, db: Session = Depends(get_db)):
    """Reset playback progress for a plan."""
    plan = crud.reset_plan_progress(db, plan_id)
    if not plan:
        raise HTTPException(status_code=404, detail="Plan not found")

    return {"message": "Progress reset", "plan_id": plan_id}

File: trunplay-backend/internal/api/plans.go
package api

import (
	"net/http"
	"strconv"

	"github.com/gin-gonic/gin"
	"github.com/trunplay/trunplay-backend/internal/db"
)

func registerPlans(r *gin.RouterGroup) {
	g := r.Group("/plans")
	g.GET("", listPlans)
	g.GET("/:plan_id", getPlan)
	g.POST("", createPlan)
	g.PUT("/:plan_id", updatePlan)
	g.DELETE("/:plan_id", deletePlan)
	g.POST("/:plan_id/activate", activatePlan)
	g.POST("/:plan_id/deactivate", deactivatePlan)
	g.POST("/:plan_id/play", playPlan)
	g.POST("/:plan_id/reset-progress", resetPlanProgress)
}

func listPlans(c *gin.Context) {
	skip, _ := strconv.Atoi(c.DefaultQuery("skip", "0"))
	limit, _ := strconv.Atoi(c.DefaultQuery("limit", "100"))
	if limit <= 0 || limit > 500 {
		limit = 100
	}
	database := getDB(c)
	tx, err := database.Begin()
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	defer tx.Rollback()
	plans, err := db.GetPlans(tx, skip, limit)
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	_ = tx.Commit()
	c.JSON(http.StatusOK, plans)
}

func getPlan(c *gin.Context) {
	planID := c.Param("plan_id")
	database := getDB(c)
	tx, err := database.Begin()
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	defer tx.Rollback()
	plan, err := db.GetPlan(tx, planID)
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	if plan == nil {
		NotFound(c, "Plan not found")
		return
	}
	_ = tx.Commit()
	c.JSON(http.StatusOK, plan)
}

func createPlan(c *gin.Context) {
	var body struct {
		Title        string  `json:"title"`
		StartTime    string  `json:"start_time"`
		EndTime      string  `json:"end_time"`
		RepeatDays   string  `json:"repeat_days"`
		SkipHolidays bool    `json:"skip_holidays"`
		DeviceID     *string `json:"device_id"`
		MediaURL     string  `json:"media_url"`
		IsActive     bool    `json:"is_active"`
		PlayMode     string  `json:"play_mode"`
	}
	if err := c.ShouldBindJSON(&body); err != nil {
		BadRequest(c, err.Error())
		return
	}
	deviceID := ""
	if body.DeviceID != nil {
		deviceID = *body.DeviceID
	}
	if body.PlayMode == "" {
		body.PlayMode = "SEQUENTIAL"
	}
	database := getDB(c)
	tx, err := database.Begin()
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	defer tx.Rollback()
	plan, err := db.CreatePlan(tx, body.Title, body.StartTime, body.EndTime, body.RepeatDays, deviceID, body.MediaURL, body.PlayMode, body.SkipHolidays, body.IsActive)
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	if plan != nil && plan["is_active"].(bool) {
		sched := getScheduler(c)
		sched.SchedulePlan(plan["id"].(string), plan["title"].(string), body.StartTime, body.RepeatDays, body.SkipHolidays)
	}
	_ = tx.Commit()
	c.JSON(http.StatusOK, plan)
}

func updatePlan(c *gin.Context) {
	planID := c.Param("plan_id")
	var body map[string]interface{}
	if err := c.ShouldBindJSON(&body); err != nil {
		BadRequest(c, err.Error())
		return
	}
	database := getDB(c)
	tx, err := database.Begin()
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	defer tx.Rollback()
	plan, err := db.UpdatePlan(tx, planID, body)
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	if plan == nil {
		NotFound(c, "Plan not found")
		return
	}
	sched := getScheduler(c)
	if plan["is_active"].(bool) {
		sched.SchedulePlan(plan["id"].(string), plan["title"].(string), plan["start_time"].(string), plan["repeat_days"].(string), plan["skip_holidays"].(bool))
	} else {
		sched.CancelPlan(planID)
	}
	_ = tx.Commit()
	c.JSON(http.StatusOK, plan)
}

func deletePlan(c *gin.Context) {
	planID := c.Param("plan_id")
	getScheduler(c).CancelPlan(planID)
	database := getDB(c)
	tx, err := database.Begin()
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	defer tx.Rollback()
	ok, err := db.DeletePlan(tx, planID)
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	if !ok {
		NotFound(c, "Plan not found")
		return
	}
	_ = tx.Commit()
	c.JSON(http.StatusOK, gin.H{"message": "Plan deleted"})
}

func activatePlan(c *gin.Context) {
	planID := c.Param("plan_id")
	database := getDB(c)
	tx, err := database.Begin()
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	defer tx.Rollback()
	plan, err := db.GetPlan(tx, planID)
	if err != nil || plan == nil {
		NotFound(c, "Plan not found")
		return
	}
	plan, err = db.UpdatePlan(tx, planID, map[string]interface{}{"is_active": true})
	if err != nil || plan == nil {
		NotFound(c, "Plan not found")
		return
	}
	getScheduler(c).SchedulePlan(plan["id"].(string), plan["title"].(string), plan["start_time"].(string), plan["repeat_days"].(string), plan["skip_holidays"].(bool))
	_ = tx.Commit()
	c.JSON(http.StatusOK, plan)
}

func deactivatePlan(c *gin.Context) {
	planID := c.Param("plan_id")
	database := getDB(c)
	tx, err := database.Begin()
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	defer tx.Rollback()
	plan, err := db.UpdatePlan(tx, planID, map[string]interface{}{"is_active": false})
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	if plan == nil {
		NotFound(c, "Plan not found")
		return
	}
	getScheduler(c).CancelPlan(planID)
	_ = tx.Commit()
	c.JSON(http.StatusOK, plan)
}

func playPlan(c *gin.Context) {
	planID := c.Param("plan_id")
	var body struct {
		StartPosition int `json:"start_position"`
	}
	_ = c.ShouldBindJSON(&body)
	database := getDB(c)
	tx, err := database.Begin()
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	defer tx.Rollback()
	plan, err := db.GetPlan(tx, planID)
	if err != nil || plan == nil {
		NotFound(c, "Plan not found")
		return
	}
	_ = tx.Commit()
	tx2, _ := database.Begin()
	success, err := getPlayback(c).StartPlayback(c.Request.Context(), tx2, plan, "MANUAL", body.StartPosition)
	if err != nil {
		tx2.Rollback()
		ServerError(c, err.Error())
		return
	}
	_ = tx2.Commit()
	if !success {
		ServerError(c, "Failed to start playback")
		return
	}
	c.JSON(http.StatusOK, gin.H{"message": "Playback started", "plan_id": planID})
}

func resetPlanProgress(c *gin.Context) {
	planID := c.Param("plan_id")
	database := getDB(c)
	tx, err := database.Begin()
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	defer tx.Rollback()
	plan, err := db.ResetPlanProgress(tx, planID)
	if err != nil {
		ServerError(c, err.Error())
		return
	}
	if plan == nil {
		NotFound(c, "Plan not found")
		return
	}
	_ = tx.Commit()
	c.JSON(http.StatusOK, gin.H{"message": "Progress reset", "plan_id": planID})
}
